mountpoint-s3/tests/fuse_tests/unlink_test.rs

```diff
@@ -0,0 +1,121 @@
+            .collect::<Vec<_>>(),
+        vec!["foo.txt", "hello.txt"]
+    );
+
+    let err = fs::remove_file(nonexistent_path).expect_err("file remove/unlink on non-existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    fs::remove_file(&path).expect("file remove/unlink of existing path should succeed");
+
+    let err = fs::remove_file(path).expect_err("file remove/unlink a second time on existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    // readdir should now show the file as gone
+    let dir = fs::read_dir(&main_dir).unwrap();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().and_then(OsStr::to_str).unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        vec!["foo.txt"]
+    );
+}
+
+#[cfg(feature = "s3_tests")]
+#[test]
+fn simple_unlink_test_s3() {
+    simple_unlink_tests(crate::fuse_tests::s3_session::new, "simple_unlink_tests");
+}
+
+#[test_case(""; "no prefix")]
+#[test_case("simple_unlink_test"; "prefix")]
+fn simple_unlink_test_mock(prefix: &str) {
+    simple_unlink_tests(crate::fuse_tests::mock_session::new, prefix);
+}
+
+/// Testing behavior when a file is unlinked in the middle of reading
+fn unlink_filehandle_test<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, Default::default());
+
+    // Add a file directly to the bucket
+    const B_IN_MB: usize = 1024 * 1024;
+    test_client.put_object("dir/this.txt", &[0u8; B_IN_MB * 128]).unwrap();
+    test_client.put_object("dir/other.txt", &[0u8; 1024]).unwrap(); // Persist implicit directory for test
+
+    let main_dir = mount_point.path().join("dir");
+    let path = mount_point.path().join("dir/this.txt");
+
+    let mut f = File::options()
+        .read(true)
+        .write(false)
+        .open(&path)
+        .expect("open should work");
+    f.read_exact(&mut [0u8; 1]).expect("read should work");
+
+    fs::remove_file(&path).expect("file remove/unlink of existing path should succeed");
+
+    // readdir should show the file as gone
+    let dir = fs::read_dir(main_dir).unwrap();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().and_then(OsStr::to_str).unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        vec!["other.txt"]
+    );
+
+    let _new_pos = f.seek(SeekFrom::Start((B_IN_MB * 120) as u64)).unwrap(); // Seek far ahead in file, to exceed prefetcher's progress
+    let err = f
+        .read_exact(&mut [0u8; 1])
+        .expect_err("fresh read using open file handle should fail as object no longer exists");
+    let raw_os_err = err.raw_os_error().expect("err should be OS-level err");
+    assert_eq!(raw_os_err, libc::EIO);
+}
```

Comment:
Is there a way we can test concurrent cases like when a file is being written locally, when we are trying to unlink.

Reply:
Oh! I was asking because I needed it for my rmdir implementation as well. Like to test the case where there is mknod/mkdir call concurrently with rmdir, and its executed after rmdir.